src/local.rs

```diff
@@ -318,58 +316,109 @@ impl ObjectStore for LocalFileSystem {
                     Ok(entry @ Some(_)) => entry
                         .filter(|dir_entry| dir_entry.file_type().is_file())
                         .map(|entry| {
-                            let location = self.filesystem_to_path(entry.path())?;
+                            let location = config.filesystem_to_path(entry.path())?;
                             convert_entry(entry, location)
                         }),
                 }
             });
 
-        Ok(stream::iter(s).boxed())
+        // If no tokio context, return iterator
+        if tokio::runtime::Handle::try_current().is_err() {
+            return Ok(futures::stream::iter(s).boxed());
+        }
+
+        // Otherwise list in batches of CHUNK_SIZE
+        const CHUNK_SIZE: usize = 1024;
+
+        let buffer = VecDeque::with_capacity(CHUNK_SIZE);
+        let stream = futures::stream::try_unfold((s, buffer), |mut state| async move {
+            if state.1.is_empty() {
```

Comment:
I would find this easier to read if instead of `|mut state|` it was `|(mut s, mut buffer)|` -- it took me some time to figure out that `buffer` was still used, but it had been renamed